src/main/java/com/github/princesslana/eriscasper/data/resource/AuditLogChange.java

```diff
@@ -0,0 +1,163 @@
+package com.github.princesslana.eriscasper.data.resource;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.github.princesslana.eriscasper.data.Data;
+import com.github.princesslana.eriscasper.data.DataException;
+import com.github.princesslana.eriscasper.data.Snowflake;
+import com.google.common.collect.ImmutableList;
+import java.util.Optional;
+
+@SuppressWarnings("unused")
+public class AuditLogChange<T> {
+
+  private Class<T> valueClass;
+  private T newValue;
+  private T oldValue;
+  private String key;
+
+  private AuditLogChange(Class<T> valueClass, T newValue, T oldValue, String key) {
+    this.valueClass = valueClass;
+    this.newValue = newValue;
+    this.oldValue = oldValue;
+    this.key = key;
+  }
+
+  public Class<T> getValueClass() {
+    return valueClass;
+  }
+
+  public boolean isArray() {
+    return valueClass == null;
+  }
+
+  public Optional<T> getNewValue() {
+    return Optional.ofNullable(newValue);
+  }
+
+  public Optional<T> getOldValue() {
+    return Optional.ofNullable(oldValue);
+  }
+
+  public String getKey() {
+    return key;
+  }
+
+  public static AuditLogChange<?> parseJson(JsonNode node) {
+    String key = node.get("key").textValue();
+    switch (key.toLowerCase()) {
+      case "$add":
+      case "$remove":
+        return newArrayInstance(node, Role.class, key);
+      case "permission_overwrites":
+        return newArrayInstance(node, Overwrite.class, key);
+      case "owner_id":
+      case "afk_channel_id":
+      case "widget_channel_id":
+      case "system_channel_id":
+      case "channel_id":
+      case "inviter_id":
+      case "id":
+      case "application_id":
+        return newInstance(node, Snowflake.class, key);
+      case "widget_enabled":
+      case "nsfw":
+      case "temporary":
+      case "deaf":
+      case "mute":
+      case "hoist":
+      case "mentionable":
+        return newInstance(node, Boolean.class, key);
+      case "permissions":
+      case "color":
+      case "allow":
+      case "deny":
+      case "max_uses":
+      case "uses":
+      case "max_age":
+      case "afk_timeout":
+      case "mfa_level":
+      case "verification_level":
+      case "explicit_content_filter":
+      case "default_message_notifications":
+      case "position":
+      case "bitrate":
+      case "prune_delete_days":
+        return newInstance(node, Integer.class, key);
+      case "type":
+      case "nick":
+      case "avatar_hash":
+      case "name":
+      case "icon_hash":
+      case "splash_hash":
+      case "code":
+      case "region":
+      case "vanity_url_code":
+      case "topic":
+        return newInstance(node, String.class, key);
+      default:
+        System.out.println("Reached here with " + key);
+        System.out.println("Body?=>" + node.asText());
```

Comment:
No System.out.println. If we want to log, we use slf4j.

This looks like a case for an exception though

Reply:
Alright, I'll throw an exception there.